02-peer-protocol.md

```diff
@@ -445,6 +446,177 @@ to broadcast the commitment transaction to get his funds back and open a new
+   * [`u32`:`funding_feerate_perkw`]
+   * [`point`:`funding_pubkey`]
+
+1. type: 76 (`splice_ack`)
+2. data:
+   * [`chain_hash`:`chain_hash`]
+   * [`channel_id`:`channel_id`]
+   * [`point`:`funding_pubkey`]
+
+#### Requirements
+
+The sender of `splice`:
+- MUST have successfully initiated quiescence.
+- MUST NOT send `splice` before sending and receiving `funding_locked`.
+- MUST NOT send `splice` while a splice is being negotiated.
+- If any splice is in progress:
+  - MUST NOT send a splice message with `funding_feerate_perkw` which is less than 1.25 the previous `funding_feerate_perkw` (rounded down).
+
+The receiver of `splice`:
+- SHOULD fail the connection if there is an ongoing splice, and the `funding_feerate_perkw` is not at least 1.25 the previous `funding_feerate_perkw` (rounded down).
+- MUST respond with `splice_ack` containing its own `funding_pubkey`.
+- MUST begin splice negotiation.
+
+The receiver of `splice_ack`:
+- MUST begin splice negotiation.
+
+
+#### Rationale
+
+There is no harm in agreeing to a splice with a high feerate
+(presumably the recipient will not contribute to the splice which they
+consider to be overpaying).
+
+## Splice Negotiation
+
+The splice negotiation is very similar to the `init_rbf` negotiation:
+both sides alternate sending `tx_add_input` and `tx_add_output` until
+they both send consecutive `tx_complete`.
+
+### Requirements
+
+(Note that many of these messages have their own, additional
+requirements detailed elsewhere).
+
+The initiator:
+- MUST `tx_add_input` an input which spends the current funding transaction output.
+- MUST `tx_add_output` a zero-value output which pays to the two funding keys using the higher of the two `generation` fields.
+- MUST pay for the common fields.
+
+Upon receipt of consecutive `tx_complete`s, each node:
+- MUST fail negotiation if there is not exactly one input spending the current funding transaction.
+- MUST fail negotiation if there is not exactly one output with zero value paying to the two funding keys (a.k.a. the new channel funding output)
+- MUST calculate the channel capacity for each side:
+  - Start with the previous balance
+  - Add that side's new inputs (excluding the one spending the current funding transaction)
+  - Subtracting each sides new outputs (except the zero-value one paying to the funding keys)
+  - Subtract the total fee that side is paying for the splice transaction.
+- MUST replace the zero-value funding output amount with the total channel capacity.
+- MUST calculate the channel balance for each side:
+  - Subtract any outstanding HTLCs offered by that side.
+- if either side has added an output other than the new channel funding output:
+  - MUST fail the negotiation if the balance for that side is less than 1% of the total channel capacity.
+- SHOULD NOT fail if the splice transaction is nonstandard.
+- MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.
```

Comment:
In order to prevent awkward states requiring force closures, double spending inputs, etc:
* The first sender of `commitment_signed` must be the same as the first sender of `tx_signatures`
* The second sending of `commitment_signed` commits that sender to be unable to `tx_abort`
* At this point both sides should have data committed to DB such that they can resume signing after a reconnect
* The first sender can now safely send `tx_signatures` or `tx_abort`
* After this point neither side can use `tx_abort`

This is needed to handle cases where `tx_signatures` is sent but not received. The sender is left in a vulnerable position where their peer may be able to broadcast the tx but they cannot. This leaves them forced to search the blockchain for the splice forever, have another successful splice, double spend an input, or force close the channel. Even if they find the splice onchain the recourse in that case would be to force close causing awkward competing transactions.

As such `tx_abort` after the signature is sent is a hostile act and shouldn't be allowed per the spec. Since `tx_signatures` can't be guaranteed to be received after being sent, the final `commitment_signed` message acts as an ack commiting the peer to signing.

I believe the correct response to receiving an invalid `tx_abort` is to:
1) Force close
2) Wait for force close to complete, while,
3) Also monitoring for the splice confirmation onchain, and,
4) Force closing that splice confirmation if it appears onchain

** Allowed examples **

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
           <- tx-sig
```

```
commit-sig -> 
           <- commit-sig
tx-abort   ->
```

** Disallowed examples **

```
commit-sig -> 
           <- commit-sig
           <- tx-abort
```

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
           <- tx-abort
```

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
           <- tx-sig
tx-abort   ->
```

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
tx-abort   ->
```

Reply:
I don't understand why we need to increment the `commitment_number`, this requires exchanging a lot of data that is completely unnecessary:

- we need to re-sign every other active commitment, but they didn't change
- we need to exchange `revoke_and_ack` for the previous commitment, but there's no good reason for that since there was no change

It looks like we add this unnecessary complexity only to be able to re-use the existing `channel_reestablish` fields (`next_commitment_number` / `next_revocation_number`) to figure out that a splice's `commit_sig` needs to be retransmitted when reconnecting.

I think it's much simpler to send a single `commit_sig` with the current `commitment_number` for the splice transaction. The only drawback is that if we get disconnected, we need to use a different mechanism to figure out whether that `commit_sig` must be retransmitted. But that mechanism should be really simple, as only the following scenarios are possible:

- there is no ongoing splice -> nothing new to retransmit
- there is an ongoing splice: peers only need to tell their counterparty whether they received their splice `commit_sig` or not (since we're in `stfu` mode, there cannot be "normal" operations to retransmit, the commitment numbers will be up-to-date on both sides)

So it feels like the addition of a single boolean to `channel_reestablish` lets us trivially deal with the reestablish logic.

> The first sender of commitment_signed must be the same as the first sender of tx_signatures
> The second sending of commitment_signed commits that sender to be unable to tx_abort
> At this point both sides should have data committed to DB such that they can resume signing after a reconnect

As detailed [here](https://gist.github.com/t-bast/2594ecf0314f0b91293a3f5e3bf10540), I think peers MUST store the channel state right after _sending_ `commit_sig`, otherwise there are cases where one side has stored the channel and the other hasn't, which leads to an unnecessary force-close. Once peers do that, we don't care about the ordering of `commit_sig`, I don't think we need to add any requirement for that.